common/cuda_hip/base/device_matrix_data_kernels.cpp

```diff
@@ -31,9 +31,13 @@ void remove_zeros(std::shared_ptr<const DefaultExecutor> exec,
     auto value_ptr = as_device_type(values.get_const_data());
     auto size = values.get_size();
     // count nonzeros
-    auto nnz = thrust::count_if(
-        thrust_policy(exec), value_ptr, value_ptr + size,
-        [] __device__(device_value_type value) { return is_nonzero(value); });
+    // __half != is only device, can not call __device__ from a __host__
+    // __device__ (is_nonzero)
```

Comment:
is this resolved with `constexpr`?

Reply:
no. constexpr does not just need marked from #1695 actually.
It still lead to __host__ tries to call __device__ function, which leads an error